### PR TITLE
api: Optimize multipart upload memory usage for unknown sized stream

### LINKED ISSUE #730: optimalPartInfo() default partSize results in running out of memory
Like #404, I'm streaming data which could be over 1GB and trying to use PutObjectStreaming(), but this results in optimalPartInfo(-1) being called and a partSize of 603979776 being returned.

The problem I face is that this partSize is used as, essentially, the size of the read buffer, and I'm running out of memory (I end up with errors like `fork/exec ...: cannot allocate memory` after the stream completes).

I've tried setting partSize = minPartSize in optimalPartInfo() when the input size is -1, and this solves my memory issue.

Regardless of exactly how it's done, can we get (or choose) a reasonable sized read buffer during PutObjectStreaming() so that we can stream in constant small amount of memory?

### PR DESCRIPTION
**Pull request by harshavardhana**
Fixes #730

### CODE REVIEW

**Comment by harshavardhana** [comment]
Any reviews here @balamurugana @vadmeste ?

**Comment by harshavardhana** [comment]
Changes are done @balamurugana - ready to review again.

File: api-put-object-multipart.go [hunk 1]
```diff
@@ -100,53 +107,54 @@ func (c Client) putObjectMultipartNoStream(bucketName, objectName string, reader
 		// HTTPS connection.
 		hashAlgos, hashSums := c.hashMaterials()
 
-		// Calculates hash sums while copying partSize bytes into tmpBuffer.
-		prtSize, rErr := hashCopyN(hashAlgos, hashSums, tmpBuffer, reader, partSize)
-		if rErr != nil && rErr != io.EOF {
+		bufp := bufPool.Get().(*[]byte)
+		prtSize, rErr := io.ReadFull(reader, *bufp)
```
Comment (balamurugana) [inline api-put-object-multipart.go]: `length` may be better than `prtSize`

File: api-put-object-multipart.go [hunk 2]
```diff
@@ -100,53 +107,54 @@ func (c Client) putObjectMultipartNoStream(bucketName, objectName string, reader
...
-
 		// Save successfully uploaded size.
-		totalUploadedSize += prtSize
+		totalUploadedSize += int64(prtSize)
 
 		// Increment part number.
 		partNumber++
 
+		// Put back data into bufpool.
+		bufPool.Put(bufp)
```
Comment (balamurugana) [inline api-put-object-multipart.go]: this statement can be in `defer`

Comment (harshavardhana) [inline api-put-object-multipart.go]: We cannot defer @balamurugana we need to get this back in the next loop.

File: api-put-object-multipart.go (hunk 1 above)
Comment (harshavardhana) [inline api-put-object-multipart.go]: Sure.

**Comment by harshavardhana** [comment]
@vadmeste need your reviews as well.

**Pull request merged**